fastfield_codecs/src/compact_space/build_compact_space.rs

```diff
@@ -0,0 +1,242 @@
+            blanks.push(blank_range);
+        }
+    };
+    for (first, second) in values_sorted.iter().tuple_windows() {
+        // Correctness Overflow: the values are deduped and sorted (BTreeSet property), that means
+        // there's always space between two values.
+        let blank_range = first + 1..=second - 1;
+        add_range(blank_range);
+    }
+
+    // Replace after stabilization of https://github.com/rust-lang/rust/issues/62924
+    // Add preceeding range if values don't start at 0
+    if let Some(first_val) = values_sorted.iter().next() {
+        if *first_val != 0 {
+            let blank_range = 0..=first_val - 1;
+            add_range(blank_range);
+        }
+    }
+
+    // Add succeeding range if values don't end at u128::MAX
+    if let Some(last_val) = values_sorted.iter().last() {
+        if *last_val != u128::MAX {
+            let blank_range = last_val + 1..=u128::MAX;
+            add_range(blank_range);
+        }
+    }
+    blanks
+}
+
+struct BlankCollector {
+    blanks: Vec<BlankRange>,
+    staged_blanks_sum: u128,
+}
+impl BlankCollector {
+    fn new() -> Self {
+        Self {
+            blanks: vec![],
+            staged_blanks_sum: 0,
+        }
+    }
+    fn stage_blank(&mut self, blank: BlankRange) {
+        self.staged_blanks_sum += blank.blank_size();
+        self.blanks.push(blank);
+    }
+    fn drain(&mut self) -> impl Iterator<Item = BlankRange> + '_ {
+        self.staged_blanks_sum = 0;
+        self.blanks.drain(..)
+    }
+    fn staged_blanks_sum(&self) -> u128 {
+        self.staged_blanks_sum
+    }
+    fn num_blanks(&self) -> usize {
+        self.blanks.len()
+    }
+}
+fn num_bits(val: u128) -> u8 {
+    (128u32 - val.leading_zeros()) as u8
+}
+
+/// Will collect blanks and add them to compact space if more bits are saved than cost from
+/// metadata.
+pub fn get_compact_space(
+    values_deduped_sorted: &BTreeSet<u128>,
+    total_num_values: usize,
+    cost_per_blank: usize,
+) -> CompactSpace {
+    let mut blanks = get_blanks(values_deduped_sorted);
+    let mut amplitude_compact_space = u128::MAX;
+    let mut amplitude_bits: u8 = num_bits(amplitude_compact_space);
+
+    let mut compact_space = CompactSpaceBuilder::new();
+    if values_deduped_sorted.is_empty() {
+        return compact_space.finish();
+    }
+
+    let mut blank_collector = BlankCollector::new();
+    // We will stage blanks until they reduce the compact space by 1 bit.
+    // Binary heap to process the gaps by their size
+    while let Some(blank_range) = blanks.pop() {
+        blank_collector.stage_blank(blank_range);
+
+        let staged_spaces_sum: u128 = blank_collector.staged_blanks_sum();
+        // +1 for later added null value
+        let amplitude_new_compact_space = amplitude_compact_space - staged_spaces_sum + 1;
+        let amplitude_new_bits = num_bits(amplitude_new_compact_space);
+        if amplitude_bits == amplitude_new_bits {
+            continue;
```

Comment:
That's tricky... 
There is ladder effect that makes it so that it is sometimes beneficial to add a blank even if it is not changing the amplitude.

Reply:
There's a difference between staging and adding blanks. We stage blanks until they are have an positive effect and then add them.

If `num_bits(amplitude)` didn't change we continue staging blanks. There's no effect on the compression yet.

If one of the continue is triggered, we don't add the staged blanks but keep staging more blanks.

Reply:
ah got it!